maestrowf/datastructures/core/executiongraph.py

```diff
@@ -906,12 +920,32 @@ def execute_ready_steps(self):
             # Now, we need to take the min of the length of the queue and the
             # computed number of slots. We could have free slots, but have less
             # in the queue.
-            _available = min(_available, len(self.ready_steps))
+            # _available = min(_available, len(self.ready_steps))  # Kind of inconsistent if it's affected by queue size sometimes
             LOGGER.info("Found %d available slots...", _available)
 
-        for i in range(0, _available):
+        # Sort the ready steps by weight (higher weights execute first)
+        pprint("STEP ORDER")
+        # que_steps = []
+        # while not self.ready_steps.empty():
+        #     que_steps.append(self.ready_steps.get())
+        #     # pprint(f"{que_steps[-1]}")
+
+        # for item in que_steps:
+        #     pprint(item)
+
+        
+        # for step_order, step in enumerate(self.ready_steps):
+        #     pprint(f"{step_order}: {step[0]}")
+        new_steps_running = 0
+        LOGGER.debug("Available slots: %d, Ready steps is empty %s",
+                     _available, self.ready_steps.empty() == True)
+        while not self.ready_steps.empty() and (new_steps_running < _available or _available < 0):
```

Comment:
This changes the original intent of the throttle that was added. The `--throttle` option meant that Maestro would not exceed that number of jobs in total. This form of throttle says that we are allowed to launch that number of new steps instead of being a global limit to my understanding. Am I incorrect there?

Also, this conditional is a little convoluted -- since `available` could be `-1` it adds the extra conditional.

Reply:
yeah, this part needs another data structure to really work right again: queue's size method is only approximate, so isn't reliable for this like the previous implementation was.   will need a running steps tracker alongside this to make this throttling less convoluted

Reply:
so, this actually does work as intended, no need to do my edit.  The controls come from `_available` getting set using throttle and the number of running steps a little above this loop.  So if `in_progress` is already = throttle this conditional will not let any new jobs through.  

just tested it on one of the hello world specs and verified it would properly limit to 1 job at a time (throttle setting) with the slurm adapter

Will add a comment to explain the variable.